homeassistant/components/mastodon/services.py

```diff
@@ -0,0 +1,148 @@
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="not_loaded",
+            translation_placeholders={"target": entry.title},
+        )
+    return cast(MastodonConfigEntry, entry)
+
+
+def setup_services(hass: HomeAssistant) -> None:
+    """Set up the services for the Mastodon integration."""
+
+    async def async_post(call: ServiceCall) -> ServiceResponse:
+        """Post a status."""
+        entry = async_get_entry(hass, call.data[ATTR_CONFIG_ENTRY_ID])
+        client = entry.runtime_data.client
+
+        status = call.data[ATTR_STATUS]
+
+        visibility: str | None = (
+            StatusVisibility(call.data[ATTR_VISIBILITY])
+            if ATTR_VISIBILITY in call.data
+            else None
+        )
+        content_warning: str | None = call.data.get(ATTR_CONTENT_WARNING)
+        media: str | None = call.data.get(ATTR_MEDIA)
+        media_warning: str | None = call.data.get(ATTR_MEDIA_WARNING)
+
+        if media:
+            if not hass.config.is_allowed_path(media):
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN,
+                    translation_key="not_whitelisted_directory",
+                    translation_placeholders={"media": media},
+                )
+            mediadata = await _upload_media(client, media)
+
+            try:
+                await hass.async_add_executor_job(
+                    partial(
+                        client.status_post,
+                        status=status,
+                        visibility=visibility,
+                        spoiler_text=content_warning,
+                        media_ids=mediadata["id"],
+                        sensitive=media_warning,
+                    )
+                )
+            except MastodonAPIError as err:
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN,
+                    translation_key="unable_to_send_message",
+                ) from err
+        else:
+            try:
+                await hass.async_add_executor_job(
+                    partial(
+                        client.status_post,
+                        status=status,
+                        visibility=visibility,
+                        spoiler_text=content_warning,
+                    )
+                )
+            except MastodonAPIError as err:
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN,
+                    translation_key="unable_to_send_message",
+                ) from err
+        return None
+
+    async def _upload_media(client: Mastodon, media_path: Any = None) -> Any:
+        """Upload media."""
+
+        media_type = get_media_type(media_path)
+        try:
+            mediadata = await hass.async_add_executor_job(
+                partial(client.media_post, media_file=media_path, mime_type=media_type)
+            )
+
```

Comment:
Would it make sense to put everything in 1 executor job instead of 2?

Switching from event loop to executor job is expensive, and in this case it does not matter _that_ much, but we still consider it a good practice to make sure we don't do 2 calls where it could be just 1

Reply:
Moving to kwargs makes this possible so done both.